File: socai/media/image.py
"""Image download, OCR, and current-backend vision helpers."""

from __future__ import annotations

import asyncio
import base64
import hashlib
import time
from typing import Any

from socai.agent.backends import Backend

from .common import (
    MediaConfig,
    MediaUnavailable,
    detect_media_type,
    download_bytes,
    save_bytes,
    short,
    url_suffix,
)
from .timing import TimingRecord


class ImageProcessor:
    def __init__(
        self,
        config: MediaConfig,
        *,
        backend: Backend | None = None,
        timing: TimingRecord | None = None,
    ):
        self.config = config
        self.backend = backend
        self.timing = timing

    def _record(self, op: str, t0: float) -> None:
        if self.timing is not None:
            self.timing.record(op, time.perf_counter() - t0)

    def download_bytes(self, url: str, *, referer: str = "") -> bytes:
        t0 = time.perf_counter()
        try:
            return download_bytes(url, referer=referer, timeout=self.config.request_timeout_s)
        finally:
            self._record("image_download", t0)

    def save_bytes(self, payload: bytes, *, label: str, suffix: str = ".bin") -> str:
        return save_bytes(self.config.base_dir, payload, label=label, suffix=suffix)

    def ocr_image(self, payload: bytes) -> str:
        if not self.config.use_ocr:
            raise MediaUnavailable("OCR is disabled")
        if not payload:
            return ""
        t0 = time.perf_counter()
        try:
            try:
                import Foundation  # type: ignore
                import Vision  # type: ignore
            except Exception as exc:  # noqa: BLE001 - optional dependency
                raise MediaUnavailable("Apple Vision OCR is unavailable; install PyObjC Vision bindings") from exc

            data = Foundation.NSData.dataWithBytes_length_(payload, len(payload))
            handler = Vision.VNImageRequestHandler.alloc().initWithData_options_(data, {})
            request = Vision.VNRecognizeTextRequest.alloc().init()
            try:
                request.setRecognitionLanguages_(["zh-Hans", "zh-Hant", "en-US"])
                request.setRecognitionLevel_(Vision.VNRequestTextRecognitionLevelAccurate)
            except Exception:
                pass
            ok, error = handler.performRequests_error_([request], None)
            if not ok:
                raise RuntimeError(f"Apple Vision OCR failed: {error}")
            lines: list[str] = []
            for observation in request.results() or []:
                try:
                    candidates = observation.topCandidates_(1)
                    if candidates:
                        text = str(candidates[0].string() or "").strip()
                        if text:
                            lines.append(text)
                except Exception:
                    continue
            return "\n".join(lines)
        finally:
            self._record("ocr_image", t0)

    def describe_image(self, payload: bytes, prompt: str, *, max_tokens: int = 180) -> str:
        if not self.config.use_vision:
            raise MediaUnavailable("Vision is disabled")
        if self.backend is None:
            raise MediaUnavailable("No agent LLM backend was provided for image vision")
        if not payload:
            return ""

        t0 = time.perf_counter()
        try:
            media_type = detect_media_type(payload) or "image/jpeg"
            response = self.backend.create_message(
                system="You describe images concisely and only state visible evidence.",
                messages=[
                    {
                        "role": "user",
                        "content": [
                            {"type": "text", "text": prompt},
                            {
                                "type": "image",
                                "source": {
                                    "type": "base64",
                                    "media_type": media_type,
                                    "data": base64.b64encode(payload).decode("ascii"),
                                },
                            },
                        ],
                    }
                ],
                tools=[],
                max_tokens=max_tokens,
            )
            return "\n".join(response.text_blocks).strip()
        finally:
            self._record("vision_image", t0)

    async def enrich_images_async(
        self,
        images: list[dict[str, Any]],
        *,
        referer: str = "",
        label: str = "image",
        run_vision: bool = False,
        vision_concurrency: int | None = None,
    ) -> list[dict[str, Any]]:
        if not images:
            return []

        concurrency = max(1, int(vision_concurrency or self.config.vision_concurrency or 1))

        # Parallel downloads.
        t_batch = time.perf_counter()
        download_results = await asyncio.gather(
            *[
                asyncio.to_thread(self._safe_download, str(image.get("url") or ""), referer)
                for image in images
            ],
            return_exceptions=False,
        )
        if self.timing is not None:
            self.timing.record("image_download_batch", time.perf_counter() - t_batch)

        # Dedup + persist.
        deduped: list[tuple[dict[str, Any], bytes]] = []
        seen: set[str] = set()
        for image, (payload, error) in zip(images, download_results):
            if not str(image.get("url") or "").strip():
                continue
            item = dict(image)
            if error:
                item["download_error"] = error
                deduped.append((item, b""))
                continue
            if not payload:
                continue
            digest = hashlib.md5(payload).hexdigest()
            if digest in seen:
                continue
            seen.add(digest)
            item["local_path"] = self.save_bytes(
                payload,
                label=f"{label}_{len(deduped) + 1}",
                suffix=url_suffix(str(image.get("url") or ""), ".jpg"),
            )
            deduped.append((item, payload))

        # Concurrent OCR + vision (vision rate-limited by semaphore).
        semaphore = asyncio.Semaphore(concurrency)

        async def enrich_one(item: dict[str, Any], payload: bytes) -> None:
            if not payload:
                return
            if self.config.use_ocr and not item.get("ocr_text"):
                try:
                    ocr_text = await asyncio.to_thread(self.ocr_image, payload)
                    if ocr_text and ocr_text.strip():
                        item["ocr_text"] = short(ocr_text, 800)
                except Exception as exc:  # noqa: BLE001 - optional capability
                    item["ocr_error"] = str(exc)
            if run_vision and self.config.use_vision and not item.get("vision_description"):
                async with semaphore:
                    try:
                        item["vision_description"] = await asyncio.to_thread(
                            self.describe_image,
                            payload,
                            "Describe this Xiaohongshu image for the note. Focus on concrete visible facts.",
                        )
                    except Exception as exc:  # noqa: BLE001 - optional capability
                        item["vision_error"] = str(exc)

        t_enrich = time.perf_counter()
        await asyncio.gather(
            *[enrich_one(item, payload) for item, payload in deduped],
            return_exceptions=False,
        )
        if self.timing is not None:
            self.timing.record("image_enrich_batch", time.perf_counter() - t_enrich)

        return [item for item, _ in deduped]

    def _safe_download(self, url: str, referer: str) -> tuple[bytes, str]:
        if not url:
            return b"", ""
        try:
            return self.download_bytes(url, referer=referer), ""
        except Exception as exc:  # noqa: BLE001 - per-media best effort
            return b"", f"{type(exc).__name__}: {exc}"

    def diagnostics(self) -> dict[str, Any]:
        return {
            "apple_vision_ocr": self._module_importable("Vision") and self._module_importable("Foundation"),
            "agent_vision_backend": type(self.backend).__name__ if self.backend else "",
        }

    @staticmethod
    def _module_importable(name: str) -> bool:
        try:
            __import__(name)
            return True
        except Exception:
            return False

File: crates/media/src/image.rs
use std::collections::HashSet;
use std::time::Instant;

use anyhow::Result;
use base64::Engine;
use serde_json::Value;
use socai_agent::{Block, Message, MessageRole, ToolSchema};

use crate::common::{detect_media_type, insert_string, short, url_suffix, MediaUnavailable};
use crate::md5;
use crate::processor::MediaProcessor;

impl MediaProcessor {
    pub fn ocr_image(&self, _payload: &[u8]) -> Result<String> {
        if !self.config.use_ocr {
            anyhow::bail!(MediaUnavailable("OCR is disabled".into()));
        }
        anyhow::bail!(MediaUnavailable(
            "OCR is unavailable in the Rust media processor".into()
        ));
    }

    pub async fn describe_image(
        &self,
        payload: &[u8],
        prompt: &str,
        max_tokens: u32,
    ) -> Result<String> {
        if !self.config.use_vision {
            anyhow::bail!(MediaUnavailable("Vision is disabled".into()));
        }
        let Some(llm_provider) = &self.llm_provider else {
            anyhow::bail!(MediaUnavailable(
                "No agent LLM provider was provided for image vision".into()
            ));
        };
        if payload.is_empty() {
            return Ok(String::new());
        }
        let t0 = Instant::now();
        let media_type = detect_media_type(payload);
        let data = base64::engine::general_purpose::STANDARD.encode(payload);
        let response = llm_provider
            .send(
                "You describe images concisely and only state visible evidence.",
                &[Message {
                    role: MessageRole::User,
                    content: socai_agent::MessageContent::Blocks(vec![
                        Block::Text {
                            text: prompt.to_string(),
                        },
                        Block::Image { data, media_type },
                    ]),
                }],
                &[] as &[ToolSchema],
                max_tokens,
            )
            .await?;
        self.timing.record("vision_image", t0.elapsed());
        Ok(response.text_blocks.join("\n").trim().to_string())
    }

    pub async fn enrich_images(
        &self,
        images: &[Value],
        referer: &str,
        label: &str,
        run_vision: bool,
    ) -> Vec<Value> {
        if images.is_empty() {
            return Vec::new();
        }

        let t_batch = Instant::now();
        let downloads = images
            .iter()
            .map(|image| {
                let url = image
                    .get("url")
                    .and_then(Value::as_str)
                    .unwrap_or("")
                    .to_string();
                self.safe_download(url, referer.to_string())
            })
            .collect::<Vec<_>>();
        let download_results = futures::future::join_all(downloads).await;
        self.timing
            .record("image_download_batch", t_batch.elapsed());

        let mut seen = HashSet::new();
        let mut deduped: Vec<(Value, Vec<u8>)> = Vec::new();
        for (image, (payload, error)) in images.iter().zip(download_results) {
            let url = image
                .get("url")
                .and_then(Value::as_str)
                .unwrap_or("")
                .trim();
            if url.is_empty() {
                continue;
            }
            let mut item = image.clone();
            if let Some(error) = error {
                insert_string(&mut item, "download_error", error);
                deduped.push((item, Vec::new()));
                continue;
            }
            if payload.is_empty() {
                continue;
            }
            let digest = md5::md5_hex(&payload);
            if !seen.insert(digest) {
                continue;
            }
            let path = self.save_bytes(
                &payload,
                &format!("{label}_{}", deduped.len() + 1),
                &url_suffix(url, ".jpg"),
            );
            match path {
                Ok(path) => insert_string(&mut item, "local_path", path.to_string_lossy()),
                Err(err) => insert_string(&mut item, "save_error", format!("{err:#}")),
            }
            deduped.push((item, payload));
        }

        let t_enrich = Instant::now();
        let mut out = Vec::with_capacity(deduped.len());
        for (mut item, payload) in deduped {
            if !payload.is_empty() && self.config.use_ocr && item.get("ocr_text").is_none() {
                match self.ocr_image(&payload) {
                    Ok(text) if !text.trim().is_empty() => {
                        insert_string(&mut item, "ocr_text", short(&text, 800));
                    }
                    Ok(_) => {}
                    Err(err) => insert_string(&mut item, "ocr_error", format!("{err:#}")),
                }
            }
            if run_vision
                && !payload.is_empty()
                && self.config.use_vision
                && item.get("vision_description").is_none()
            {
                match self
                    .describe_image(
                        &payload,
                        "Describe this Xiaohongshu image for the note. Focus on concrete visible facts.",
                        180,
                    )
                    .await
                {
                    Ok(text) if !text.trim().is_empty() => {
                        insert_string(&mut item, "vision_description", text);
                    }
                    Ok(_) => {}
                    Err(err) => insert_string(&mut item, "vision_error", format!("{err:#}")),
                }
            }
            out.push(item);
        }
        self.timing.record("image_enrich_batch", t_enrich.elapsed());
        out
    }

    async fn safe_download(&self, url: String, referer: String) -> (Vec<u8>, Option<String>) {
        if url.trim().is_empty() {
            return (Vec::new(), None);
        }
        match self.download_bytes(&url, &referer).await {
            Ok(bytes) => (bytes, None),
            Err(err) => (Vec::new(), Some(format!("{err:#}"))),
        }
    }
}
